src/validation.rs

```diff
@@ -0,0 +1,141 @@
+/// This module provides functionality to validate (lint) existing module and assembly files,
+/// to check if the files meet the template structure and other requirements.
+
+use std::fmt;
+use std::fs;
+use std::path::Path;
+// use std::process::exit;
+use log::{debug, error, info, warn};
+use regex::{Regex, RegexBuilder};
+
+use crate::module::ModuleType;
+
+const ASSEMBLY_TESTS: [IssueDefinition; 2] = [
+    // Test that an assembly includes no other assemblies
+    IssueDefinition {
+        pattern: r"^include::.*assembly[_-].*\.adoc",
+        description: "This assembly includes another assembly.",
+        severity: IssueSeverity::Error,
+    },
+    // Test that files don't use the unsupported leveloffset configuration
+    IssueDefinition {
+        pattern: r"^:leveloffset:\s*\+\d*",
+        description: "Unsupported level offset configuration.",
+        severity: IssueSeverity::Error,
+    },
+];
+
+#[derive(Clone, Copy, Debug)]
+struct IssueDefinition {
+    pattern: &'static str,
+    description: &'static str,
+    severity: IssueSeverity,
+}
+
+#[derive(Clone, Copy, Debug)]
+enum IssueSeverity {
+    Information,
+    Warning,
+    Error,
+}
+
+#[derive(Debug)]
+struct IssueReport {
+    line_number: i32,
+    description: &'static str,
+    severity: IssueSeverity,
+}
+
+impl fmt::Display for IssueReport {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Line {}: {}", self.line_number, self.description)
+    }
+}
+
+pub fn validate(file_name: &str) {
+    debug!("Validating file `{}`", file_name);
+
+    let path = Path::new(file_name);
+    let base_name = path.file_name().unwrap().to_str().unwrap();
+
+    let read_result = fs::read_to_string(path);
+    let content = match read_result {
+        Ok(content) => content,
+        Err(err) => {
+            error!("Error reading file `{}`: {}", file_name, err);
+            return;
+        }
+    };
+
+    let mod_type = determine_mod_type(base_name, &content);
+
+    if mod_type == Some(ModuleType::Assembly) {
+        assembly_tests(base_name, &content);
+    }
+}
+
+fn determine_mod_type(base_name: &str, content: &str) -> Option<ModuleType> {
+    // An inner function to encapsulate the logic.
+    // This can't be a simple expression block because the logic involves
+    // stuff like for loops.
+    let inner = || {
+        if base_name.starts_with("assembly_") || base_name.starts_with("assembly-") {
+            return Some(ModuleType::Assembly);
```

Comment:
Are all assemblies called `assembly[_-].*` ?

Reply:
Well, some projects use modules with no prefixes, and that probably applies to assemblies as well. So there can conceivably be assembly files without the `assembly[_-].*` prefix.

The way that the validation detects the module type needs to be much more sophisticated than it is right now, anyway. It has to deal with cases where a file looks like one type in some places but another type in other places, and report that as an error.

For assemblies, my current idea is that if a file has no prefix and it includes other modules, `newdoc` should report "Please check if this is really an assembly".

Reply:
Ah, so it's a first version? I'm so used to draft PRs that I thought this is "finished"...

Reply:
Absolutely, this is a work in progress, an early one. See the "WIP" in the title :-)

I'll let you know when it's finished, if you'd like to take a look then.